src/writer.rs

```diff
@@ -0,0 +1,373 @@
+        writer.write_f32::<B>(*f).unwrap();
+    }
+    writer.write_all(&header.intent_name).unwrap();
+    writer.write_all(&header.magic).unwrap();
+
+    // Empty 4 bytes after the header
+    writer.write_u32::<B>(0).unwrap();
+
+    // We finally write the data.
+    // Like NiBabel, we iterate by "slice" to improve speed and use less memory
+
+    // Need the transpose for fortran used in nifti file format.
+    let data = data.t();
+
+    let arr_len = data.subview(Axis(0), 0).len();
+    let resolution: usize = header.dim[1..(header.dim[0] + 1) as usize]
+        .iter().map(|d| *d as usize).product();
+    let nb_bytes = resolution * header.bitpix as usize / 8;
+    let dim0_size = data.shape()[0];
+
+    // `1.0x + 0.0` would give the same results, but we avoid a lot of divisions
+    let slope = if header.scl_slope == 0.0 { 1.0 } else { header.scl_slope };
+    if slope != 1.0 || header.scl_inter != 0.0 {
+        let slope = A::from_f32(slope).unwrap();
+        let inter = A::from_f32(header.scl_inter).unwrap();
+        for arr_data in data.axis_iter(Axis(0)) {
+            let arr_data = arr_data.sub(inter).div(slope);
+            let arr_data = arr_data.into_shape(arr_len).unwrap();
+            let arr_data = arr_data.as_slice().unwrap();
+
+            let buffer = to_bytes(&arr_data, nb_bytes / dim0_size);
+            writer.write_all(&buffer).unwrap();
+        }
+    } else {
+        for arr_data in data.axis_iter(Axis(0)) {
+            // We need to own the data because of the into_shape() for `C` ordering.
+            let arr_data = arr_data.to_owned();
+            let arr_data = arr_data.into_shape(arr_len).unwrap();
+            let arr_data = arr_data.as_slice().unwrap();
+
+            let buffer = to_bytes(&arr_data, nb_bytes / dim0_size);
+            writer.write_all(&buffer).unwrap();
+        }
+    }
+}
+
+fn to_bytes<T>(data: &[T], size: usize) -> &[u8] {
+    unsafe { from_raw_parts::<u8>(data.as_ptr() as *const u8, size) }
+}
+
+#[cfg(test)]
+pub mod tests {
+    extern crate tempfile;
+
+    use super::*;
+
+    use ndarray::{Array, Array2, Ix2, IxDyn, ShapeBuilder};
+    use self::tempfile::tempdir;
+
+    use {InMemNiftiObject, IntoNdArray, header::MAGIC_CODE_NIP1, object::NiftiObject};
+
+    fn get_random_path(ext: &str) -> String {
+        let dir = tempdir().unwrap();
+        let path = if ext == "" {
+            dir.into_path()
+        } else {
+            dir.into_path().join(ext)
+        };
+        path.to_str().unwrap().to_string()
+    }
+
+    pub fn generate_nifti_header(
+        dim: [u16; 8],
+        scl_slope: f32,
+        scl_inter: f32,
+        datatype: i16
+    ) -> NiftiHeader {
+        let bitpix = (mem::size_of::<f32>() * 8) as i16;
+        let magic = *MAGIC_CODE_NIP1;
+        NiftiHeader {
+            dim, datatype, bitpix, magic, scl_slope, scl_inter, ..NiftiHeader::default()
+        }
+    }
+
+    fn read_2d_image(path: &str) -> Array2<f32> {
```

Comment:
The usual pattern for taking a file path is `P where P: AsRef<Path>`.

```suggestion
    fn read_2d_image<P>(path: P) -> Array2<f32>
    where
        P: AsRef<::std::path::Path>,
    {
```